pkg/policy/selectorcache.go

```diff
@@ -808,15 +845,19 @@ func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api
 		sc.selectors[key] = newFQDNSel
 	}
 
-	// Allocate identities corresponding to the slice of IPs identified as
-	// being selected by this FQDN selector above. This could plausibly
-	// happen twice, once with an empty 'ids' slice and once with the real
-	// 'ids' slice. Either way, they are added to the selector that is
-	// stored in 'sc.selectors[]'.
-	newFQDNSel.allocateIdentityMappings(sc.idAllocator, selectorIPMapping)
+	identitiesToRelease = newFQDNSel.transferIdentityReferencesToSelector(currentlyAllocatedIdentities)
 	newFQDNSel.updateSelections()
 
-	return newFQDNSel, newFQDNSel.addUser(user)
+	return newFQDNSel, newFQDNSel.addUser(user), identitiesToRelease
+}
+
+// AddFQDNSelector adds the given api.FQDNSelector in to the selector cache. If
+// an identical EndpointSelector has already been cached, the corresponding
+// CachedSelector is returned, otherwise one is created and added to the cache.
+func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool) {
+	cs, added, identitiesToRelease := sc.addFQDNSelector(user, fqdnSelec)
+	sc.releaseIdentityMappings(identitiesToRelease)
```

Comment:
I would be slightly happier if these identities were released inside `addFQDNSelector()` rather than returning them. It would require to use an explicit `sc.mutex.Unlock()` instead of deferring it, though.

Reply:
:+1: Rolling in this diff on top:

```
$ git diff origin/submit/fqdn-deadlock-fix..HEAD
diff --git a/pkg/policy/selectorcache.go b/pkg/policy/selectorcache.go
index dccbba1366f6..9de5203e17fd 100644
--- a/pkg/policy/selectorcache.go
+++ b/pkg/policy/selectorcache.go
@@ -776,7 +776,10 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
        return identitiesToRelease
 }
 
-func (sc *SelectorCache) addFQDNSelector(user CachedSelectionUser, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool, identitiesToRelease []identity.NumericIdentity) {
+// AddFQDNSelector adds the given api.FQDNSelector in to the selector cache. If
+// an identical EndpointSelector has already been cached, the corresponding
+// CachedSelector is returned, otherwise one is created and added to the cache.
+func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool) {
        key := fqdnSelec.String()
 
        // Lock NameManager before the SelectorCache
@@ -789,7 +792,7 @@ func (sc *SelectorCache) addFQDNSelector(user CachedSelectionUser, fqdnSelec api
        if exists {
                added := fqdnSel.addUser(user)
                sc.mutex.Unlock()
-               return fqdnSel, added, nil
+               return fqdnSel, added
        }
        sc.mutex.Unlock()
 
@@ -834,8 +837,6 @@ func (sc *SelectorCache) addFQDNSelector(user CachedSelectionUser, fqdnSelec api
        // is already cached, or is a new one).
 
        sc.mutex.Lock()
-       defer sc.mutex.Unlock()
-
        // Check whether the selectorCache was updated while 'newFQDNSel' was
        // being registered without the 'sc.mutex'. If so, use it. Otherwise
        // we can use the one we just created/configured above.
@@ -844,20 +845,14 @@ func (sc *SelectorCache) addFQDNSelector(user CachedSelectionUser, fqdnSelec api
        } else {
                sc.selectors[key] = newFQDNSel
        }
-
-       identitiesToRelease = newFQDNSel.transferIdentityReferencesToSelector(currentlyAllocatedIdentities)
+       identitiesToRelease := newFQDNSel.transferIdentityReferencesToSelector(currentlyAllocatedIdentities)
        newFQDNSel.updateSelections()
+       added = newFQDNSel.addUser(user)
+       sc.mutex.Unlock()
 
-       return newFQDNSel, newFQDNSel.addUser(user), identitiesToRelease
-}
-
-// AddFQDNSelector adds the given api.FQDNSelector in to the selector cache. If
-// an identical EndpointSelector has already been cached, the corresponding
-// CachedSelector is returned, otherwise one is created and added to the cache.
-func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool) {
-       cs, added, identitiesToRelease := sc.addFQDNSelector(user, fqdnSelec)
        sc.releaseIdentityMappings(identitiesToRelease)
-       return cs, added
+
+       return newFQDNSel, added
 }
 
 // FindCachedIdentitySelector finds the given api.EndpointSelector in the
```